### PR TITLE
[ADP-3300] Fix `chainSyncFetchNextBlock`

### PR DESCRIPTION
**Pull request by HeinrichApfelmus**
This pull request fixes the implementation of `chainSyncFetchNextBlock`.

It turns out that the `cardano-node` sends `RollBackward` after successfully finding an intersection. This message typically helps when implementing a client — but I failed to consider it here. Fixed now.

### Issue number

ADP-3300

### CODE REVIEW

File: lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs [hunk 1]
```diff
@@ -266,29 +266,34 @@ chainSyncFetchNextBlock queue =
...
+        :: point -> ClientStIdle block point tip m Void
+    clientRequestBlock point =
+        SendMsgRequestNext (clientStNext point) (pure $ clientStNext point)
 
     -- Fetch the block on rollforward.
     clientStNext
-        :: ClientStNext block point tip m Void
-    clientStNext = ClientStNext
+        :: point -> ClientStNext block point tip m Void
+    clientStNext point = ClientStNext
```
Comment (paweljakubas) [inline lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs:287]: ehh, this `point` is traveling a lot downwards ;-) but that is ok here ....

Comment (HeinrichApfelmus) [reply lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs:287]: 🛷

File: lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs [hunk 2]
```diff
@@ -266,29 +266,34 @@ chainSyncFetchNextBlock queue =
...
             result $ Just block
-        , recvMsgRollBackward = \_ _ ->
-            result Nothing
+        , recvMsgRollBackward = \point' _ ->
+            -- Note: When starting the chainsync protocol,
+            -- the cardano-node may send an initial rollback message,
+            -- which we will detect here and treat as a no-op.
+            if point' == point
+            then ChainSyncClient $ pure $ clientRequestBlock point
+            else result Nothing
```
Comment (paweljakubas) [inline lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs:296]: should that be switched? when `point /= point'` then we sync with requesting the block?

Comment (HeinrichApfelmus) [reply lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs:296]: No, this is correct: when there is a rollback from `point` to the same `point'`, we ignore the rollback. When the rollback is to someplace else, this means that the `point` is no longer part of the consensus chain, and we need to abort.

Remember, this function is not intended to run the ChainSync protocol properly — instead, it uses ChainSync to fetch one particular block (namely the one after the `point`).

Comment (paweljakubas) [reply lib/network-layer/src/Cardano/Wallet/Network/Implementation/Ouroboros.hs:296]: aha, thanks! That is correct, indeed!

**Review by paweljakubas — APPROVED** [review]
:+1:

**Pull request merged**